Support GHC 8.6.1

GHC 8.6.1 was [released](https://mail.haskell.org/pipermail/glasgow-haskell-users/2018-September/026803.html). Using this version of GHC on upstream, I'm getting the following error:

```bash
$ cabal install
[1 of 2] Compiling Data.Equivalence.STT ( src/Data/Equivalence/STT.hs, dist/build/Data/Equivalence/STT.o )

src/Data/Equivalence/STT.hs:253:5: error:
    • Could not deduce (Control.Monad.Fail.MonadFail (STT s m))
        arising from a do statement
        with the failable pattern ‘dy@Root {entryWeight = wy,
                                            entryDesc = chy, entryValue = vy}’
      from the context: (Monad m, Applicative m, Ord a)
        bound by the type signature for:
                   equateEntry :: forall (m :: * -> *) a s c.
                                  (Monad m, Applicative m, Ord a) =>
                                  Equiv s c a -> Entry s c a -> Entry s c a -> STT s m (Entry s c a)
        at src/Data/Equivalence/STT.hs:249:1-116
    • In a stmt of a 'do' block:
        dy@Root {entryWeight = wy, entryDesc = chy,
                 entryValue = vy} <- readSTRef ry
      In the expression:
        do dx@Root {entryWeight = wx, entryDesc = chx,
                    entryValue = vx} <- readSTRef rx
           dy@Root {entryWeight = wy, entryDesc = chy,
                    entryValue = vy} <- readSTRef ry
           if wx >= wy then
               do writeSTRef ry ...
                  ....
           else
               do writeSTRef rx ...
                  ....
      In the expression:
        if (rx /= ry) then
            do dx@Root {entryWeight = wx, entryDesc = chx,
                        entryValue = vx} <- readSTRef rx
               dy@Root {entryWeight = wy, entryDesc = chy,
                        entryValue = vy} <- readSTRef ry
               if wx >= wy then do ... else do ...
        else
            return repx
    |
253 |     dy@Root{entryWeight = wy, entryDesc = chy, entryValue = vy} <- readSTRef ry
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Failed to install equivalence-0.3.2
```

Blocking https://github.com/agda/agda/issues/3160.

--- a/src/Data/Equivalence/STT.hs
+++ b/src/Data/Equivalence/STT.hs
@@ -249,20 +249,23 @@ equivalence class -}
 equateEntry :: (Monad m, Applicative m, Ord a) => Equiv s c a -> Entry s c a -> Entry s c a -> STT s m (Entry s c a)
 equateEntry Equiv {combDesc = mkDesc} repx@(Entry rx) repy@(Entry ry) = 
   if (rx /= ry) then do
-    dx@Root{entryWeight = wx, entryDesc = chx, entryValue = vx} <- readSTRef rx
-    dy@Root{entryWeight = wy, entryDesc = chy, entryValue = vy} <- readSTRef ry
-    if  wx >= wy
-      then do
-        writeSTRef ry Node {entryParent = repx, entryValue = vy}
-        writeSTRef rx dx{entryWeight = wx + wy, entryDesc = mkDesc chx chy}
-        return repx
-      else do
-       writeSTRef rx Node {entryParent = repy, entryValue = vx}
-       writeSTRef ry dy{entryWeight = wx + wy, entryDesc = mkDesc chx chy}
-       return repy
-    else return  repx
+    dx <- readSTRef rx
+    dy <- readSTRef ry
+    case (dx, dy) of
+      ( Root{entryWeight = wx, entryDesc = chx, entryValue = vx}
+        , Root{entryWeight = wy, entryDesc = chy, entryValue = vy} ) ->
+        if  wx >= wy
+        then do
+          writeSTRef ry Node {entryParent = repx, entryValue = vy}
+          writeSTRef rx dx{entryWeight = wx + wy, entryDesc = mkDesc chx chy}
+          return repx
+        else do
+          writeSTRef rx Node {entryParent = repy, entryValue = vx}
+          writeSTRef ry dy{entryWeight = wx + wy, entryDesc = mkDesc chx chy}
+          return repy
 
-
+      _ -> fail "error on `equateEntry`"
+  else return  repx
 
 combineEntries :: (Monad m, Applicative m, Ord a)
                => Equiv s c a -> [b] -> (b -> STT s m (Entry s c a)) -> STT s m ()

